File: services/shared/schemas/draft.ts
import { z } from "zod";

// A draft is the canonical writeable artifact a content_factory crew produces.
// Per-platform reshapes carry parentDraftId.

export const DraftStatusSchema = z.enum([
  "drafting",
  "in_review",
  "awaiting_approval",
  "approved",
  "scheduled",
  "published",
  "denied",
  "archived",
]);
export type DraftStatus = z.infer<typeof DraftStatusSchema>;

export const ChannelSchema = z.enum([
  "x",
  "linkedin",
  "reddit",
  "tiktok",
  "instagram",
  "newsletter",
  "blog",
]);
export type Channel = z.infer<typeof ChannelSchema>;

export const ClaimSchema = z.object({
  statement: z.string().min(1),
  supportingUrl: z.string().url().optional(),
  snippet: z.string().optional(),
  retrievedAt: z.string().datetime().optional(),
});
export type Claim = z.infer<typeof ClaimSchema>;

export const DraftSchema = z.object({
  id: z.string().uuid(),
  companyId: z.string().uuid(),
  clientId: z.string().uuid().nullable().optional(),
  parentDraftId: z.string().uuid().nullable().optional(), // long-form -> per-channel reshape
  channel: ChannelSchema,
  status: DraftStatusSchema.default("drafting"),
  title: z.string().max(300).optional(),
  body: z.string(),
  hashtags: z.array(z.string()).default([]),
  claims: z.array(ClaimSchema).default([]),
  voiceScore: z.number().min(0).max(1).nullable().optional(),
  predictedPercentile: z.number().min(0).max(100).nullable().optional(),
  authoredByAgentId: z.string().uuid().nullable().optional(),
  approvalId: z.string().uuid().nullable().optional(),
  scheduledAt: z.string().datetime().nullable().optional(),
  publishedAt: z.string().datetime().nullable().optional(),
  publishedUrl: z.string().url().nullable().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});
export type Draft = z.infer<typeof DraftSchema>;

File: services/shared/schemas/draft.py
"""Draft schema — Python mirror of draft.ts."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

DraftStatus = Literal[
    "drafting",
    "in_review",
    "awaiting_approval",
    "approved",
    "scheduled",
    "published",
    "denied",
    "archived",
]

Channel = Literal["x", "linkedin", "reddit", "tiktok", "instagram", "newsletter", "blog"]


class Claim(BaseModel):
    statement: str = Field(..., min_length=1)
    supporting_url: str | None = None
    snippet: str | None = None
    retrieved_at: datetime | None = None


class Draft(BaseModel):
    id: str
    company_id: str
    client_id: str | None = None
    parent_draft_id: str | None = None
    channel: Channel
    status: DraftStatus = "drafting"
    title: str | None = Field(default=None, max_length=300)
    body: str
    hashtags: list[str] = Field(default_factory=list)
    claims: list[Claim] = Field(default_factory=list)
    voice_score: float | None = Field(default=None, ge=0.0, le=1.0)
    predicted_percentile: float | None = Field(default=None, ge=0.0, le=100.0)
    authored_by_agent_id: str | None = None
    approval_id: str | None = None
    scheduled_at: datetime | None = None
    published_at: datetime | None = None
    published_url: str | None = None
    created_at: datetime
    updated_at: datetime
